File: cpp/03_thread_pool.cpp
/*
PROBLEM: Thread Pool Implementation

Implement a reusable thread pool that manages a fixed number of worker threads
and distributes tasks to them. This is a fundamental building block for
concurrent applications.

REQUIREMENTS:
- Create a fixed pool of N worker threads
- Implement thread-safe task queue
- Provide submit() method that returns a future for result
- Support exception handling through futures
- Provide shutdown() method for graceful termination
- Use std::future and std::promise for result handling

PERFORMANCE NOTES:
- Should efficiently distribute work among threads
- Threads should not busy-wait (use condition variables)
- Should handle 10000+ tasks efficiently
- Minimal overhead per task submission

TEST CASE EXPECTATIONS:
- Submit 1000 tasks to pool with 4 workers, all should complete
- Tasks should execute in parallel (measure with timing)
- Futures should return results correctly
- Exceptions in tasks should propagate through futures
- Shutdown should wait for all tasks to complete
*/

#include <iostream>
#include <thread>
#include <queue>
#include <mutex>
#include <condition_variable>
#include <future>
#include <functional>
#include <vector>
#include <cassert>
#include <chrono>
#include <memory>

class ThreadPool {
public:
    ThreadPool(size_t num_workers = 4) : num_workers_(num_workers), shutdown_(false) {
        // TODO: Create and start worker threads
    }

    ~ThreadPool() {
        // TODO: Ensure proper shutdown
    }

    // Submit a task to the pool and return a future for the result
    template <typename F, typename... Args>
    auto submit(F&& f, Args&&... args) -> std::future<std::invoke_result_t<F, Args...>> {
        using return_type = std::invoke_result_t<F, Args...>;

        // TODO: Create a packaged_task with the function
        // TODO: Get a future from the packaged_task
        // TODO: Wrap the task for the queue
        // TODO: Add to task queue
        // TODO: Notify worker
        // TODO: Return future

        // Placeholder return
        std::promise<return_type> p;
        return p.get_future();
    }

    // Shutdown the thread pool
    // wait: if true, wait for all tasks to complete
    void shutdown(bool wait = true) {
        // TODO: Signal workers to stop
        // TODO: If wait, join all threads
        // TODO: If not wait, detach threads or force exit
    }

    size_t pending_tasks() const {
        // TODO: Return number of queued tasks
        return 0;
    }

private:
    void worker_loop() {
        // TODO: Implement worker thread main loop
        // - Get tasks from queue
        // - Execute them
        // - Handle shutdown signal
        // - Exit when shutdown and queue is empty
    }

    size_t num_workers_;
    bool shutdown_;
    std::queue<std::function<void()>> task_queue_;
    std::mutex queue_mutex_;
    std::condition_variable queue_cv_;
    std::vector<std::thread> workers_;
};


int simple_task(int x, int y = 2) {
    // Simulate some work
    std::this_thread::sleep_for(std::chrono::milliseconds(10));
    return x * y;
}

void test_basic_submit() {
    ThreadPool pool(2);

    std::vector<std::future<int>> futures;

    for (int i = 0; i < 5; ++i) {
        auto future = pool.submit(simple_task, i, 2);
        futures.push_back(std::move(future));
    }

    std::vector<int> results;
    for (auto& future : futures) {
        results.push_back(future.get());
    }

    std::vector<int> expected = {0, 2, 4, 6, 8};
    assert(results == expected);

    pool.shutdown(true);
    std::cout << "✓ Basic submit test passed (" << results.size() << " tasks)\n";
}


void test_parallel_execution() {
    ThreadPool pool(4);

    std::vector<std::future<int>> futures;
    const int num_tasks = 10;

    auto start = std::chrono::high_resolution_clock::now();

    for (int i = 0; i < num_tasks; ++i) {
        auto future = pool.submit(simple_task, i);
        futures.push_back(std::move(future));
    }

    // Wait for all to complete
    for (auto& future : futures) {
        future.get();
    }

    auto end = std::chrono::high_resolution_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end - start);

    pool.shutdown(false);

    // With 4 workers and 10 tasks at 10ms each, should take ~30ms (not 100ms)
    // Allow for overhead
    int expected_sequential_ms = num_tasks * 10;
    assert(duration.count() < expected_sequential_ms / 3);

    std::cout << "✓ Parallel execution test passed (10 tasks in " << duration.count() << "ms)\n";
}


void test_exception_handling() {
    ThreadPool pool(2);

    auto failing_task = []() -> int {
        throw std::runtime_error("Task failed");
    };

    auto future = pool.submit(failing_task);

    try {
        future.get();
        assert(false); // Should have thrown
    } catch (const std::runtime_error& e) {
        assert(std::string(e.what()) == "Task failed");
    }

    pool.shutdown(true);
    std::cout << "✓ Exception handling test passed\n";
}


void test_multiple_tasks() {
    ThreadPool pool(4);

    std::vector<std::future<int>> futures;
    const int num_tasks = 100;

    for (int i = 0; i < num_tasks; ++i) {
        auto future = pool.submit(simple_task, i % 10);
        futures.push_back(std::move(future));
    }

    std::vector<int> results;
    for (auto& future : futures) {
        results.push_back(future.get());
    }

    assert(results.size() == num_tasks);

    pool.shutdown(true);
    std::cout << "✓ Multiple tasks test passed (" << results.size() << " tasks)\n";
}


void test_shutdown_wait() {
    ThreadPool pool(2);

    std::vector<std::future<int>> futures;

    auto slow_task = []() {
        std::this_thread::sleep_for(std::chrono::milliseconds(100));
        return 42;
    };

    for (int i = 0; i < 5; ++i) {
        auto future = pool.submit(slow_task);
        futures.push_back(std::move(future));
    }

    auto start = std::chrono::high_resolution_clock::now();
    pool.shutdown(true);
    auto elapsed = std::chrono::duration_cast<std::chrono::milliseconds>(
        std::chrono::high_resolution_clock::now() - start
    );

    // Shutdown should have waited
    assert(elapsed.count() >= 100);

    // All futures should be ready
    for (auto& future : futures) {
        assert(future.get() == 42);
    }

    std::cout << "✓ Shutdown wait test passed (waited " << elapsed.count() << "ms)\n";
}


int main() {
    std::cout << "Running ThreadPool tests...\n\n";

    test_basic_submit();
    test_parallel_execution();
    test_exception_handling();
    test_multiple_tasks();
    test_shutdown_wait();

    std::cout << "\n✓ All tests passed!\n";

    return 0;
}

File: python/03_thread_pool.py
"""
PROBLEM: Thread Pool Executor

Implement a simple thread pool that manages a fixed number of worker threads and
distributes tasks from a queue to these workers. Workers should process tasks
and return results.

REQUIREMENTS:
- Create a fixed pool of N worker threads
- Implement a task queue for distributing work
- Worker threads should process tasks from queue
- Provide a submit() method to add tasks
- Provide a shutdown() method for graceful termination
- Track task results/completion

PERFORMANCE NOTES:
- Should efficiently distribute work among threads
- Threads should not busy-wait
- Shutdown should wait for all tasks to complete
- Should handle 1000+ tasks efficiently

TEST CASE EXPECTATIONS:
- Submit 100 tasks to pool with 4 workers, all should complete
- Tasks should execute in parallel (measure with timing)
- Results should be collectable in any order
- Shutdown should block until all tasks complete
"""

import threading
import time
from typing import Any, Callable, Optional
from queue import Queue, Empty


class Task:
    """Represents a task to be executed."""

    def __init__(self, task_id: int, func: Callable, args: tuple = (), kwargs: dict = None):
        self.task_id = task_id
        self.func = func
        self.args = args
        self.kwargs = kwargs or {}
        self.result = None
        self.exception = None
        self.completed = False
        self.event = threading.Event()

    def execute(self):
        """Execute the task and store result."""
        try:
            self.result = self.func(*self.args, **self.kwargs)
        except Exception as e:
            self.exception = e
        finally:
            self.completed = True
            self.event.set()

    def wait(self, timeout: Optional[float] = None) -> bool:
        """Wait for task to complete."""
        return self.event.wait(timeout)

    def get_result(self, timeout: Optional[float] = None) -> Any:
        """Get the result of the task."""
        if not self.wait(timeout):
            raise TimeoutError(f"Task {self.task_id} did not complete in time")
        if self.exception:
            raise self.exception
        return self.result


class ThreadPool:
    """A simple thread pool executor."""

    def __init__(self, num_workers: int = 4):
        """
        Initialize the thread pool.

        Args:
            num_workers: Number of worker threads to create
        """
        self.num_workers = num_workers
        self.task_queue = Queue()
        # TODO: Initialize workers and other necessary state
        self.workers = []
        self.running = True
        self.task_counter = 0
        self.counter_lock = threading.Lock()

    def _worker(self):
        """Worker thread main loop."""
        # TODO: Implement worker thread logic
        # - Get tasks from queue
        # - Execute them
        # - Handle shutdown signal
        pass

    def start(self):
        """Start all worker threads."""
        # TODO: Create and start worker threads
        pass

    def submit(self, func: Callable, *args, **kwargs) -> Task:
        """
        Submit a task to the pool.

        Args:
            func: Callable to execute
            args: Positional arguments for the callable
            kwargs: Keyword arguments for the callable

        Returns:
            Task object that can be used to get results
        """
        # TODO: Create task, add to queue, return task object
        pass

    def shutdown(self, wait: bool = True):
        """
        Shut down the thread pool.

        Args:
            wait: If True, wait for all tasks to complete before returning
        """
        # TODO: Signal workers to stop and optionally wait for completion
        pass


def simple_task(x: int, y: int = 2) -> int:
    """A simple task that returns x^y."""
    time.sleep(0.01)  # Simulate work
    return x ** y


def failing_task():
    """A task that raises an exception."""
    raise ValueError("This task always fails")


def test_basic_submit():
    """Test submitting basic tasks."""
    pool = ThreadPool(num_workers=2)
    pool.start()

    tasks = []
    for i in range(5):
        task = pool.submit(simple_task, i, y=2)
        tasks.append(task)

    results = [task.get_result() for task in tasks]
    expected = [i ** 2 for i in range(5)]

    assert results == expected, f"Expected {expected}, got {results}"
    pool.shutdown(wait=True)

    print(f"✓ Basic submit test passed ({len(results)} tasks)")


def test_parallel_execution():
    """Test that tasks execute in parallel."""
    pool = ThreadPool(num_workers=4)
    pool.start()

    num_tasks = 10
    start_time = time.time()

    tasks = []
    for i in range(num_tasks):
        task = pool.submit(simple_task, i)
        tasks.append(task)

    # Wait for all to complete
    for task in tasks:
        task.get_result()

    elapsed = time.time() - start_time
    pool.shutdown(wait=False)

    # With 4 workers and 10 tasks at 0.01s each, should take ~0.03s (not 0.1s)
    # Allow some overhead
    expected_sequential = num_tasks * 0.01
    assert elapsed < expected_sequential * 0.5, (
        f"Tasks don't seem to run in parallel. "
        f"Took {elapsed:.3f}s, expected ~{expected_sequential * 0.25:.3f}s"
    )

    print(f"✓ Parallel execution test passed (10 tasks in {elapsed:.3f}s)")


def test_exception_handling():
    """Test handling of exceptions in tasks."""
    pool = ThreadPool(num_workers=2)
    pool.start()

    task = pool.submit(failing_task)

    try:
        result = task.get_result()
        assert False, "Should have raised ValueError"
    except ValueError as e:
        assert str(e) == "This task always fails"

    pool.shutdown(wait=True)
    print("✓ Exception handling test passed")


def test_multiple_tasks():
    """Test submitting many tasks."""
    pool = ThreadPool(num_workers=4)
    pool.start()

    num_tasks = 100
    tasks = []

    for i in range(num_tasks):
        task = pool.submit(simple_task, i % 10)
        tasks.append(task)

    results = [task.get_result() for task in tasks]

    assert len(results) == num_tasks
    assert all(isinstance(r, int) for r in results)

    pool.shutdown(wait=True)
    print(f"✓ Multiple tasks test passed ({num_tasks} tasks)")


def test_shutdown_wait():
    """Test that shutdown waits for tasks."""
    pool = ThreadPool(num_workers=2)
    pool.start()

    slow_task_completed = []

    def slow_task(duration):
        time.sleep(duration)
        slow_task_completed.append(True)

    tasks = [pool.submit(slow_task, 0.1) for _ in range(5)]

    start = time.time()
    pool.shutdown(wait=True)
    elapsed = time.time() - start

    assert len(slow_task_completed) == 5, "Not all tasks completed"
    assert elapsed >= 0.1, "Shutdown returned too quickly"

    print(f"✓ Shutdown wait test passed (waited {elapsed:.3f}s)")


if __name__ == "__main__":
    test_basic_submit()
    test_parallel_execution()
    test_exception_handling()
    test_multiple_tasks()
    test_shutdown_wait()
    print("\n✓ All tests passed!")
